engine/src/main/java/org/kigalisim/engine/support/RetireRecalcStrategy.java

```diff
@@ -0,0 +1,164 @@
+
+  private Scope scope;
+
+  /**
+   * Create a new RetireRecalcStrategy.
+   *
+   * @param scope The scope to use for calculations, null to use engine's current scope
+   */
+  public RetireRecalcStrategy(Scope scope) {
+    this.scope = scope;
+  }
+
+  @Override
+  public void execute(Engine target) {
+    if (!(target instanceof SingleThreadEngine)) {
+      throw new IllegalArgumentException(
+          "RetireRecalcStrategy requires a SingleThreadEngine");
+    }
+
+    SingleThreadEngine engine = (SingleThreadEngine) target;
+
+    // Move the logic from SingleThreadEngine.recalcRetire
+    // Setup
+    OverridingConverterStateGetter stateGetter =
+        new OverridingConverterStateGetter(engine.getStateGetter());
+    UnitConverter unitConverter = new UnitConverter(stateGetter);
+    Scope scopeEffective = scope != null ? scope : engine.getScope();
+    String application = scopeEffective.getApplication();
+    String substance = scopeEffective.getSubstance();
+
+    // Check allowed
+    if (application == null || substance == null) {
+      engine.raiseNoAppOrSubstance("recalculating population change", "");
+    }
+
+    // Calculate change
+    EngineNumber currentPriorRaw = engine.getStreamKeeper().getStream(
+        application,
+        substance,
+        "priorEquipment"
+    );
+    EngineNumber currentPrior = unitConverter.convert(currentPriorRaw, "units");
+
+    EngineNumber currentEquipmentRaw = engine.getStreamKeeper().getStream(
+        application,
+        substance,
+        "equipment"
+    );
+    EngineNumber currentEquipment = unitConverter.convert(currentEquipmentRaw, "units");
+
+    stateGetter.setPopulation(currentPrior);
+    EngineNumber amountRaw = engine.getStreamKeeper().getRetirementRate(application, substance);
+    EngineNumber amount = unitConverter.convert(amountRaw, "units");
+    stateGetter.clearPopulation();
+
+    // Calculate new values
+    BigDecimal newPriorValue = currentPrior.getValue().subtract(amount.getValue());
+    BigDecimal newEquipmentValue = currentEquipment.getValue().subtract(amount.getValue());
+
+    EngineNumber newPrior = new EngineNumber(newPriorValue, "units");
+    EngineNumber newEquipment = new EngineNumber(newEquipmentValue, "units");
+
+    // Update equipment streams
+    engine.getStreamKeeper().setStream(application, substance, "priorEquipment", newPrior);
+    engine.getStreamKeeper().setStream(application, substance, "equipment", newEquipment);
+
+    // Update GHG accounting
+    engine.recalcEolEmissions(scopeEffective);
+
+    // Propagate
+    engine.recalcPopulationChange(null, null);
+    engine.recalcSales(null);
+    engine.recalcConsumption(null);
+  }
+
+  @Override
+  public void execute(Engine target, RecalcKit kit) {
+    // Move the logic from SingleThreadEngine.recalcRetire
+    // Setup
+    OverridingConverterStateGetter stateGetter =
+        new OverridingConverterStateGetter(kit.getStateGetter().orElseThrow(
```

Comment:
I think this is frequently repeated. Can we make getStateGetter return a realized StateGetter and have the RecalcKitBuilder throw an exception if one is not given?